app/upgrades/v11/upgrades.go

```diff
@@ -58,3 +78,80 @@ func MigrateEscrowAccounts(ctx sdk.Context, ak authkeeper.AccountKeeper) {
 		ak.SetModuleAccount(ctx, acc)
 	}
 }
+
+// HandleMainnetUpgrade handles the logic for the reward distribution, it only commits to the db if successful
+func HandleRewardDistribution(ctx sdk.Context, bk bankkeeper.Keeper, sk stakingkeeper.Keeper, dk distributionkeeper.Keeper, logger log.Logger) {
+	// use a cache context as a rollback mechanism in case
+	// the distrbution fails
+	cacheCtx, writeFn := ctx.CacheContext()
+	err := DistributeRewards(cacheCtx, bk, sk, dk)
+	if err != nil {
+		// log error instead of aborting the upgrade
+		logger.Error("failed to distribute rewards", "error", err.Error())
+	} else {
+		writeFn()
+	}
+}
+
+// DistributeRewards distributes the token allocations from the Olympus Mons incentivized testnet
+func DistributeRewards(ctx sdk.Context, bk bankkeeper.Keeper, sk stakingkeeper.Keeper, dk distributionkeeper.Keeper) error {
+	fundingAccountAddress := sdk.MustAccAddressFromBech32(FundingAccount)
+	numValidators := sdk.NewInt(int64(len(Validators)))
+
+	for _, currentDistribute := range Accounts {
+
+		// move rewards to the receiving account
+		receivingAccount := sdk.MustAccAddressFromBech32(currentDistribute[0])
+		receivingAmount, ok := sdk.NewIntFromString(currentDistribute[1])
+		if !ok {
+			return errorsmod.Wrapf(errortypes.ErrInvalidType,
+				"cannot retrieve allocation from string for address %s",
+				currentDistribute[0])
+		}
+		currentRewards := sdk.Coins{
+			sdk.NewCoin(types.BaseDenom, receivingAmount),
+		}
+		err := bk.SendCoins(ctx, fundingAccountAddress, receivingAccount, currentRewards)
+		if err != nil {
+			return err
+		}
+
+		// stake from the receiving account to all validators equally
+		currentStakeAmount := (currentRewards.QuoInt(numValidators)[0]).Amount // only one coin in slice
+		remainderAmount := (currentRewards[0].Amount).Mod(numValidators)       // same as above
+		for i, validatorBech32 := range Validators {
+			validatorAddress, err := sdk.ValAddressFromBech32(validatorBech32)
+			if err != nil {
+				return err
+			}
+			validator, found := sk.GetValidator(ctx, validatorAddress)
+			if !found {
+				return errorsmod.Wrapf(slashingtypes.ErrBadValidatorAddr,
+					"validator address %s cannot be found",
+					validatorAddress)
+			}
+			// 1 signifies unbonded tokens, subtractAccount being true means delegation, not redelegation
+			_, err = sk.Delegate(ctx, receivingAccount, currentStakeAmount, 1, validator, true)
+			if err != nil {
+				return err
+			}
+			// we delegate the remainder to the first validator, for the sake of testing consistency
+			// this remainder is on the order of 10^-18 evmos,  is at most 10^-15 evmos after all rewards allocated
+			if remainderAmount.IsPositive() && i == 0 {
+				_, err = sk.Delegate(ctx, receivingAccount, remainderAmount, 1, validator, true)
```

Comment:
why call `Delegate` twice for the first validator? Just add it to the first validators delegationAmount before performing the delegation the first time